llvm/lib/CAS/LazyMappedFileRegion.cpp

```diff
@@ -242,9 +242,59 @@ Expected<LazyMappedFileRegion> LazyMappedFileRegion::create(
 
   sys::fs::file_t File = sys::fs::convertFDToNativeFile(FD);
 
+  struct FileLockRAII {
+    LazyMappedFileRegion &LMFR;
+    bool IsLocked = false;
+
+    enum LockKind { Shared, Exclusive };
+
+    FileLockRAII(LazyMappedFileRegion &LMFR) : LMFR(LMFR) {}
+    ~FileLockRAII() {
+      if (Error E = unlock())
+        consumeError(std::move(E));
+    }
+
+    Error lock(LockKind LK) {
+      if (Error E = unlock())
```

Comment:
I just noticed this: why does locking unlock first? Typically locks are either non-recursive, in which case it's an error to lock it twice, or recursive, in which case locking it twice is okay.  But here locking twice is "okay", but it releases the lock in between, which seems unexpected.

Is this about upgrading shared to exclusive lock? If so I would suggest client does the unlock so it's explicit.

Reply:
Yeah, I would suggest to do the lock switching explicitly in the client code too. I would say make the lock twice an error.

Reply:
For unix it would be fine to just call `lock` again but there is this weird behavior of [windows file lock](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex) that "If the same range is locked with an exclusive and a shared lock, two unlock operations are necessary to unlock the region".
I understand that this piece of code is in the unix part but I wanted a pattern that could potentially work on the windows part as well (if someone investigates what is going on with that).

In any case, I can move the unlock to the caller.

Reply:
See updated commit.